csharp/src/Google.Protobuf/Collections/MapField.cs

```diff
@@ -221,17 +227,15 @@ public bool TryGetValue(TKey key, out TValue value)
             }
         }
 
-        // TODO: Make these views?
-
         /// <summary>
         /// Gets a collection containing the keys in the map.
         /// </summary>
-        public ICollection<TKey> Keys { get { return list.Select(t => t.Key).ToList(); } }
+        public ICollection<TKey> Keys { get { return new MapView<TKey>(this, pair => pair.Key, ContainsKey); } }
```

Comment:
Sg.
Actually, when thinking of this more in detail - caching the reference doesn't guarantee you will always get the object - if two threads call this and you are unlucky, both will obtain a different reference and only one of them would be cached. So, given it's impossible to guarantee only one reference will ever be returned, the expectation I was living with turns out to be wrong.

Based on this realization, I don't feel strongly about this. Feel free to choose approach that makes most sense to you.